### PR TITLE
[mpi] Fix #508

### LINKED ISSUE #508: [unstable][arrays] matrix transpose() + broadcast()
The following test fails for me when run on >1 cores.
```c++
TEST(Arrays, transpose_bcast) {

 mpi::communicator world;

 matrix<dcomplex> A = {{1, 2, 3},
                       {4, 5, 6}};
 matrix<dcomplex> At = A.transpose();

 matrix<dcomplex> B;
 if(world.rank() == 0) B = At;

 mpi_broadcast(B, world, 0);

 EXPECT_ARRAY_EQ(At, B);
}
```
(`B` ends up having the same shape as `At`, but different traversal order).

This problem apparently has something to do with  #451.

### PR DESCRIPTION
**Pull request by krivenko**
Don't know if I chose the right place for that `set_mem_layout_or_check_if_view()`  function...

### CODE REVIEW

File: triqs/arrays/impl/indexmap_storage_pair.hpp [hunk 1]
```diff
@@ -170,6 +170,7 @@ namespace triqs { namespace arrays {
 
     auto const & memory_layout() const { return indexmap_.memory_layout(); }
     indexmap_type const & indexmap() const {return indexmap_;}
+    indexmap_type & indexmap() {return indexmap_;}
```
Comment (Wentzell) [inline triqs/arrays/impl/indexmap_storage_pair.hpp]: I don't think its a good idea to expose the indexmap via a non-const reference.
The few functions that need actually change indexmap_ should be friend functions.

File: triqs/arrays/mpi.hpp [hunk 2]
```diff
@@ -70,8 +70,13 @@ namespace arrays {
   template <typename A> REQUIRES_IS_ARRAY mpi_broadcast(A &a, mpi::communicator c = {}, int root = 0) {
    if (!has_contiguous_data(a)) TRIQS_RUNTIME_ERROR << "Non contiguous view in mpi_broadcast";
    auto sh = a.shape();
+   auto m_pos = a.indexmap().memory_layout().get_memory_positions();
    MPI_Bcast(&sh[0], sh.size(), mpi::mpi_datatype<typename decltype(sh)::value_type>(), root, c.get());
-   if (c.rank() != root) resize_or_check_if_view(a, sh);
+   MPI_Bcast(&m_pos[0], m_pos.size(), mpi::mpi_datatype<typename decltype(m_pos)::value_type>(), root, c.get());
+   if (c.rank() != root) {
+     resize_or_check_if_view(a, sh);
+     set_mem_layout_or_check_if_view(a, memory_layout_t<A::rank>(m_pos));
```
Comment (Wentzell) [inline triqs/arrays/mpi.hpp]: This function should NEVER be invoked without a proper resize, am I correct?
Maybe we could just wrap the two operations into one function.

File: triqs/arrays/impl/indexmap_storage_pair.hpp (hunk 1 above)
Comment (krivenko) [reply triqs/arrays/impl/indexmap_storage_pair.hpp]: It is a bad idea, I agree, but how would you implement `set_mem_layout_or_check_if_view()` then?
Making it a friend of `array`, `matrix` and `vector` isn't a great solution either.

File: triqs/arrays/mpi.hpp (hunk 2 above)
Comment (krivenko) [reply triqs/arrays/mpi.hpp]: Do you want to call `resize_or_check_if_view()` from `set_mem_layout_or_check_if_view()` or rather call both of them from yet another function? Can you suggest a name for the combined function?

**Comment by Wentzell** [comment]
How about the following solution:

- Adjust the resize functions to take a memory layout defaulted to the one of the amv
- Adjust the ```resize_or_check_if_view``` with a default accordingly
- Get rid of set_mem_layout_or_check_if_view
- Keep only ```const &``` accessor of indexmap

If we want to possibility to ever change the memory layout after construction of an amv,
it makes sense to put it into resize in my opinion.

@parcollet What do you think?

**Comment by parcollet** [comment]
I agree with Nils.
1- indexmap should be immutable.
2- Indeed the layout must be broadcasted, so the best solution indeed seems to have
a resize overload taking both the sizes and the layout.

**Comment by krivenko** [comment]
I am trying to rewrite the fix according to your suggestions.
It is a bit difficult to overload `array & array::resize (Args && ... args)`...

**Comment by parcollet** [comment]
ok, fine.
Thanks !

**Pull request merged**